src/LanguageServer/Protocol/Handler/WorkDoneProgress/WorkDoneProgressManager.cs

```diff
@@ -0,0 +1,174 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
+
+/// <summary>
+/// Manages server initiated work done progress reporting to the client.
+/// See https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#serverInitiatedProgress
+/// </summary>
+class WorkDoneProgressManager(IClientLanguageServerManager clientLanguageServerManager) : ILspService
+{
+    private readonly IClientLanguageServerManager _clientLanguageServerManager = clientLanguageServerManager;
+
+    private class WorkDoneProgressReporter : IWorkDoneProgressReporter
```

Comment:
nit: I would find this easier to follow if inner types were all at the bottom. Don't feel strongly about it though.